Build guide: zome functions and lifecycle callbacks (#512)
holochain/docs-pages: src/pages/build/callbacks-and-lifecycle-hooks.md

Note that this can create "hot spots" where some agents have a heavier data storage and network traffic burden than others. Read the [anchors and paths section under Links, Paths, and Anchors](/build/links-paths-and-anchors/#anchors-and-paths) for more info.

!!!

This `init` callback also does something useful: it grants all peers in the network permission to send messages to an agent's [remote signal receiver callback](#define-a-recv-remote-signal-callback). (Note that this can create a risk of spamming.) {#init-grant-unrestricted-access-to-recv-remote-signal}

```rust
use hdk::prelude::*;

#[hdk_extern]
pub fn init(_: ()) -> ExternResult<InitCallbackResult> {
    let mut fns = BTreeSet::new();
    fns.insert((zome_info()?.name, "recv_remote_signal".into()));
    create_cap_grant(CapGrantEntry {
        tag: "".into(),
        access: CapAccess::Unrestricted,
        functions: GrantedFunctions::Listed(fns),
    })?;

    Ok(InitCallbackResult::Pass)
}
```

### Define a `recv_remote_signal` callback

<!-- TODO: move this to the signals page after it's written -->

> Agents in a network can send messages to each other via [remote signals](/concepts/9_signals/#remote-signals). In order to handle these signals, your coordinator zome needs to define a `recv_remote_signal` callback. Remote signals get routed from the emitting coordinator zome on the sender's machine to one of the same name on the receiver's machine.

Author: ```suggestion
Agents in a network can send messages to each other via [remote signals](/concepts/9_signals/#remote-signals). In order to handle these signals, your coordinator zome needs to define a `recv_remote_signal` callback. Remote signals get routed from the emitting coordinator zome on the sender's machine to a coordinator with the same name on the receiver's machine.
```

### Define a `validate` callback

In order to validate DHT data, you'll need to define a `validate` callback. It must take a single argument of type [`Op`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/op/enum.Op.html) and return a value of type [`ValidateCallbackResult`](https://docs.rs/hdi/latest/hdi/prelude/enum.ValidateCallbackResult.html) wrapped in an `ExternResult`.

The `validate` callback is called at two times:

1. When an agent tries to author an [action](/build/working-with-data/#entries-actions-and-records-primary-data), and
2. When an agent receives a [DHT operation](/concepts/4_dht/#a-cloud-of-witnesses) to store and serve as part of the shared database.

The nature of validation is out of scope for this page (we'll write a page on it soon), but here's a very basic example of a validation callback that approves everything: <!-- TODO: remove this example when the validation page is written -->

```rust
use hdi::prelude::*;

#[hdk_extern]
pub fn validate(_: Op) -> ExternResult<ValidateCallbackResult> {
    Ok(ValidateCallbackResult::Valid)
}
```

> ### Define a `genesis_self_check` callback

Author: @ThetaSinner how's this? hopefully less alarmist now 😌

Reviewer: Nice, yes I find this much easier to follow and I think it's clear

`init` must take an empty `()` input argument and return an [`InitCallbackResult`](https://docs.rs/hdk/latest/hdk/prelude/enum.InitCallbackResult.html) wrapped in an `ExternResult`. All zomes' `init` callbacks in a DNA must return a success result in order for cell initialization to succeed; otherwise any data written in these callbacks, along with the `InitZomesComplete` action, will be rolled back. _If any zome's init callback returns an `InitCallbackResult::Fail`, initialization will fail._ Otherwise, if any init callback returns an `InitCallbackResult::UnresolvedDependencies`, initialization will be retried at the next zome call attempt.

Here's an `init` callback that [links](/build/links-paths-and-anchors/) the [agent's ID](/build/identifiers/#agent) to the [DNA hash](/build/identifiers/#dna) as a sort of "I'm here" note. (It depends on a couple things being defined in your integrity zome; we'll show the integrity zome after this sample for completeness.)

> ```rust

Author: I struggled to find a meaningful `init` callback example; this is about as minimal as I can think of for the "link agent ID to well-known hash" use case.

use foo_integrity::{get_participant_registration_anchor, LinkTypes};
use hdk::prelude::*;

#[hdk_extern]
pub fn init(_: ()) -> ExternResult<InitCallbackResult> {
    let participant_registration_anchor_hash = get_participant_registration_anchor_hash()?;
    let AgentInfo { agent_latest_pubkey: my_pubkey, ..} = agent_info()?;
    create_link(
        participant_registration_anchor_hash,
        my_pubkey,
        LinkTypes::ParticipantRegistration,
        ()
    )?;

    Ok(InitCallbackResult::Pass)
}
```

Here's the `foo_integrity` zome code needed to make this work:

```rust
use hdi::prelude::*

#[hdk_link_types]
pub enum LinkTypes {
    ParticipantRegistration,
}

// This is a very simple implementation of the Anchor pattern, which you can
// read about in https://developer.holochain.org/build/links-paths-and-anchors/
// You don't need to tell Holochain about it with the `hdk_entry_types` macro,
// because it never gets stored -- we only use it to calculate a hash.
#[hdk_entry_helper]

> pub struct Anchor(pub Vec<u8>);

Author: I'm trying to avoid paths and anchors here, to reduce complexity, but maybe they're not as complex as all that. I'll go check the API and consider it.

Reviewer: ```
let mut hash = holo_hash::blake2b_256(url.as_str().as_bytes());
hash.extend_from_slice(&[0, 0, 0, 0]);
ExternalHash::from_raw_36(hash)
```

Can always use hashes that aren't defined as a type as a base for discovery.

The type `Anchor` is already defined in the HDI. Why does a type need to be defined here?

Author: For that matter, it could just be an `ExternalHash` whose contents are `"my hovercraft is full of eels!!".as_bytes()`, but then you need to turn the `hashing` feature flag on in `holo_hash` in order to use `.from_raw_39`. Speaking of which, I'm a bit concerned about padding the hash with `0x00000000` to supply loc bytes; will that cause problems in routing? (I understand you don't need to specify the loc bytes anymore with kitsune2)

Author: Defining a type was a misguided attempt to get away from introducing the concept of paths and anchors here, and needing to mention that you need to define a link type for the anchor and whatnot, but perhaps what I did is just as out-of-scope. Maybe I'll just byte the bullet and use the built-in stuff.

### Define a `validate` callback

In order to validate your data you'll need to define a `validate` callback. It must take a single argument of type [`Op`](https://docs.rs/holochain_integrity_types/latest/holochain_integrity_types/op/enum.Op.html) and return a value of type [`ValidateCallbackResult`](https://docs.rs/hdi/latest/hdi/prelude/enum.ValidateCallbackResult.html) wrapped in an `ExternResult`.

The `validate` callback is called at two times:

1. When an agent tries to author an [action](/build/working-with-data/#entries-actions-and-records-primary-data), and
2. When an agent receives a [DHT operation](/concepts/4_dht/#a-cloud-of-witnesses) to store and serve as part of the shared database.

The nature of validation is out of scope for this page (we'll write a page on it soon), but here's a very basic example of a validation callback that approves everything: <!-- TODO: remove this example when the validation page is written -->

```rust
use hdi::prelude::*;

#[hdk_extern]
pub fn validate(_: Op) -> ExternResult<ValidateCallbackResult> {
    Ok(ValidateCallbackResult::Valid)
}
```

> And here's an example of one that rejects everything. You'll note that the outer result is `Ok`; you should generally reserve `Err` for unexpected failures such as inability to deserialize data. However, Holochain will treat both `Ok(Invalid)` and `Err` as invalid operations that should be rejected.

Author: awesome. This claim I was feeling very shaky about. Cunningham's Law wins again.

> And here's an example of one that rejects everything. You'll note that the outer result is `Ok`; you should generally reserve `Err` for unexpected failures such as inability to deserialize data. However, Holochain will treat both `Ok(Invalid)` and `Err` as invalid operations that should be rejected.

Author: re: giving details about validation, yes, it's a complex topic, but I'm already working on a page. This is just a stub, as mentioned in a previous paragraph.

```rust
use hdi::prelude::*;

#[hdk_extern]
pub fn validate(_: Op) -> ExternResult<ValidateCallbackResult> {
    Ok(ValidateCallbackResult::Invalid("I reject everything".into()))
}
```

### Define an `entry_defs` callback

You don't need to write this callback by hand; you can let the `hdk_entry_types` macro do it for you. Read the [Define an entry type section under Entries](/build/entries/#define-an-entry-type) to find out how.

### Define a `genesis_self_check` callback

Holochain assumes that every agent is able to self-validate all the data they create before storing it in their [source chain](/concepts/3_source_chain/) and publishing it to the [DHT](/concepts/4_dht/). But at **genesis** time, when their cell has just been instantiated but they haven't connected to other peers, they may not be able to fully validate their [**genesis records**](/concepts/3_source_chain/#source-chain-your-own-data-store) if their validity depends on shared data. So Holochain skips full self-validation for these records, only validating the basic structure of their [actions](/build/working-with-data/#entries-actions-and-records-primary-data).

> This creates a risk to the new agent; they may mistakenly publish malformed data and be rejected from the network. You can define a `genesis_self_check` function that checks the _content_ of genesis records before they're published. This function is limited --- it naturally doesn't have access to DHT data. But it can be a useful guard against a [membrane proof](/resources/glossary/#membrane-proof) that the participant typed or pasted incorrectly, for example.

Author: > Like if I don't define this callback right then some of my users won't be able to join the network.

That's true, isn't it? or at least should be -- malformed validation code will be faithfully executed and the consequences faithfully applied. I didn't quite understand what you're saying in the second paragraph? There's no DNA info to validate in `GenesisSelfCheckDataV2` (although I do see it was in `V1`).

IMO `genesis_self_check` should _only_ validate the membrane proof, not the agent pubkey, and should be a strict subset of the `validate` callback's branch for `Op::RegisterAgentActivity` where the action is `Action::AgentValidationPkg`. I'll elaborate on this in the validation PR. The subset being the stuff that doesn't rely on DHT deps.

> This creates a risk to the new agent; they may mistakenly publish malformed data and be rejected from the network. You can define a `genesis_self_check` function that checks the _content_ of genesis records before they're published. This function is limited --- it naturally doesn't have access to DHT data. But it can be a useful guard against a [membrane proof](/resources/glossary/#membrane-proof) that the participant typed or pasted incorrectly, for example.

Reviewer: No that's not true at all. Holochain generates genesis records just the same for every app. If there was a problem with that then we'd spot it before we shipped. So unless you're checking a memproof, you do not need to implement this in order for your app to function.

> This creates a risk to the new agent; they may mistakenly publish malformed data and be rejected from the network. You can define a `genesis_self_check` function that checks the _content_ of genesis records before they're published. This function is limited --- it naturally doesn't have access to DHT data. But it can be a useful guard against a [membrane proof](/resources/glossary/#membrane-proof) that the participant typed or pasted incorrectly, for example.

Reviewer: I think there's also an assumption here that people are writing app validation rules for genesis. I don't believe that it's generally necessary to do anything more than sys validation does for genesis records. If the order of those ops is wrong or the agent key you're claiming isn't able to sign actions then it'll be discovered by sys validation. Most apps have no reason to implement the equivalent of a genesis self check for other peers in app validation. Again, I think that's only relevant with memproofs

> This creates a risk to the new agent; they may mistakenly publish malformed data and be rejected from the network. You can define a `genesis_self_check` function that checks the _content_ of genesis records before they're published. This function is limited --- it naturally doesn't have access to DHT data. But it can be a useful guard against a [membrane proof](/resources/glossary/#membrane-proof) that the participant typed or pasted incorrectly, for example.

Author: > No, that's not true at all. 

Which statement specifically?

Otherwise I'm a tad confused, because I'm in accord with everything you're saying but it sounds like you have concerns about things this paragraph (or my comments) is saying. These are my assumptions:

* Genesis records really don't need to be validated, _except_
* If you've got a memproof, you'll want to validate it, _but_
* If your memproof has a silly data entry mistake like a missing character, you won't find out until you join the network and get a warrant, _so_
* `genesis_self_check` protects you from a small class of silly mistakes

One thing I'm wondering is, does my text make it seem like you _have_ to care about memproofs and define a `genesis_self_check` for all DNAs? Is that the concern here?

> This creates a risk to the new agent; they may mistakenly publish malformed data and be rejected from the network. You can define a `genesis_self_check` function that checks the _content_ of genesis records before they're published. This function is limited --- it naturally doesn't have access to DHT data. But it can be a useful guard against a [membrane proof](/resources/glossary/#membrane-proof) that the participant typed or pasted incorrectly, for example.

Reviewer: I'm not suggesting that the content is wrong. I'm just suggesting that it reads as alarming when I don't think it needs to be.

This is advice to app developers so I'd prefer something worded like "If you're enforcing membrane proof rules in your `validate` callback then to let the agent validate their own joining proof, you need to do the same checks in the `genesis_self_check`. You can do all the same checks on the genesis records here."

> This creates a risk to the new agent; they may mistakenly publish malformed data and be rejected from the network

I've been trying to suggest that I read this as alarming and then you have to read about a bunch of details that I find hard to follow and only apply in some situations. I think it's clearer to give an example of when and why you'd define the genesis callback. Then the risk of having people misunderstand what the rules are and getting concerned about agents being rejected from the network is lower.

> This creates a risk to the new agent; they may mistakenly publish malformed data and be rejected from the network. You can define a `genesis_self_check` function that checks the _content_ of genesis records before they're published. This function is limited --- it naturally doesn't have access to DHT data. But it can be a useful guard against a [membrane proof](/resources/glossary/#membrane-proof) that the participant typed or pasted incorrectly, for example.

Author: Awesome, thank you for the clarification!

`genesis_self_check` must take a single argument of type [`GenesisSelfCheckData`](https://docs.rs/hdi/latest/hdi/prelude/type.GenesisSelfCheckData.html) and return a value of type [`ValidateCallbackResult`](https://docs.rs/hdi/latest/hdi/prelude/enum.ValidateCallbackResult.html) wrapped in an `ExternResult`.

Here's an example that checks that the membrane proof exists and is the right length: <!-- TODO: move this to the validation page too -->

```rust
use hdi::prelude::*;

#[hdk_extern]
pub fn genesis_self_check(data: GenesisSelfCheckData) -> ExternResult<ValidateCallbackResult> {
    if let Some(membrane_proof) = data.membrane_proof {
        if membrane_proof.bytes().len() == 32 {
            return Ok(ValidateCallbackResult::Valid);
        }
        return Ok(ValidateCallbackResult::Invalid("Membrane proof is not the right length. Please check it and enter it again.".into()));
    }
    Ok(ValidateCallbackResult::Invalid("This network needs a membrane proof to join.".into()))
}
```

## Coordinator zomes

> A [coordinator zome](/build/zomes/#coordinator) may define some lifecycle hooks: `init`, `post_commit`, and `recv_remote_signal`.

Author: I'm wondering if it makes sense to get into implementation details? I'm not explicitly calling it a lifecycle hook but it does behave like a callback, so I'm content to be a bit fudgy here.

But if there are a lot of agents in the network, this can create "hot spots" where one set of agents --- the ones responsible for storing the base address for all those links --- carry an outsized burden compared to other agents. Read the [anchors and paths section under Links, Paths, and Anchors](/build/links-paths-and-anchors/#anchors-and-paths) for more info.

!!!

This `init` callback also does something useful: it grants all peers in the network permission to send messages to an agent's [remote signal receiver callback](#define-a-recv-remote-signal-callback).

```rust
use hdk::prelude::*;

#[hdk_extern]
pub fn init(_: ()) -> ExternResult<InitCallbackResult> {
    let mut fns = BTreeSet::new();
    fns.insert((zome_info()?.name, "recv_remote_signal".into()));
    create_cap_grant(CapGrantEntry {
        tag: "".into(),
        access: CapAccess::Unrestricted,
        functions: GrantedFunctions::Listed(fns),
    })?;

    Ok(InitCallbackResult::Pass)
}
```

### Define a `recv_remote_signal` callback

<!-- TODO: move this to the signals page after it's written -->

> Agents in a network can send messages to each other via [remote signals](/concepts/9_signals/#remote-signals). In order to handle these signals, your coordinator zome needs to define a `recv_remote_signal` callback. Remote signals get routed from the emitting coordinator zome on the sender's machine to the same one on the receiver's machine, so there's no need for a coordinator to handle message types it doesn't know about.

Reviewer: > so there's no need for a coordinator to handle message types it doesn't know about.

That doesn't seem like a natural consequence of the preceding sentence. The first sentence should probably stop and this should be a new sentence, not a `,`

> Agents in a network can send messages to each other via [remote signals](/concepts/9_signals/#remote-signals). In order to handle these signals, your coordinator zome needs to define a `recv_remote_signal` callback. Remote signals get routed from the emitting coordinator zome on the sender's machine to the same one on the receiver's machine, so there's no need for a coordinator to handle message types it doesn't know about.

Author: Is it correct to say it _tries_ to dispatch to the same zome? FWIU `send_remote_signal` is roughly equivalent to

```rust
call_remote(
    agent_id,
    zome_info()?.name,
    "recv_remote_signal",
    secret,
    payload
).ok();
```

Does it feel like details about coordinator swapping are pertinent here? Happy to put that in, but also trying to be attentive to not getting into the details.

> Agents in a network can send messages to each other via [remote signals](/concepts/9_signals/#remote-signals). In order to handle these signals, your coordinator zome needs to define a `recv_remote_signal` callback. Remote signals get routed from the emitting coordinator zome on the sender's machine to the same one on the receiver's machine, so there's no need for a coordinator to handle message types it doesn't know about.

Reviewer: I don't think we should explain zome swapping. I'm advocating for specific language but not more detail So rather than "to the same one on the receiver's machine", something like "to the coordinator zome with the same role on the receiver's machine"

> Agents in a network can send messages to each other via [remote signals](/concepts/9_signals/#remote-signals). In order to handle these signals, your coordinator zome needs to define a `recv_remote_signal` callback. Remote signals get routed from the emitting coordinator zome on the sender's machine to the same one on the receiver's machine, so there's no need for a coordinator to handle message types it doesn't know about.

Author: oh, okay, I thought you were suggesting I add something about coordinator swapping. Phew.

`recv_remote_signal` takes a single argument of any type you like --- if your coordinator zome deals with multiple message types, consider creating an enum for all of them. It must return an empty `ExternResult<()>`, as this callback is not called as a result of direct interaction from the local agent and has nowhere to pass a return value.

This zome function and remote signal receiver callback implement a "heartbeat" to let everyone keep track of who's currently online. It assumes that you'll combine the two `init` callback examples in the previous section, which set up the necessary links and permissions to make this work.

```rust
use foo_integrity::LinkTypes;
use hdk::prelude::*;

// We're creating this type for both remote signals to other peers and local
// signals to the UI.
#[derive(Serialize, Deserialize, Debug)]
enum Signal {
    Heartbeat(AgentPubKey),
}

#[hdk_extern]
pub fn recv_remote_signal(payload: Signal) -> ExternResult<()> {
    if let Signal::Heartbeat(agent_id) = payload {
        // Pass the heartbeat along to my UI so it can update the other
        // peer's online status.

>         return emit_signal(Signal::Heartbeat(agent_id));

Author: interesting; what's the difference?

If you need to do any follow-up, it's safer to do this in a lifecycle hook called `post_commit`, which is called after Holochain's [call-zome workflow](/build/zome-functions/#zome-function-call-lifecycle) successfully writes its actions to the source chain. (Function calls that don't write data won't trigger this event.)

`post_commit` must take a single argument of type <code>Vec&lt;<a href="https://docs.rs/hdk/latest/hdk/prelude/type.SignedActionHashed.html">SignedActionHashed</a>&gt;</code>, which contains all the actions the function call wrote, and it must return an empty `ExternResult<()>`. This callback **must not write any data**, but it may call other zome functions in the same cell or any other local or remote cell, and it may send local or remote signals.

Here's an example that uses `post_commit` to tell someone a movie loan has been created for them. It uses the integrity zome examples from [Identifiers](/build/identifiers/#in-dht-data).

```rust
use movie_integrity::{EntryTypes, Movie, MovieLoan, UnitEntryTypes};
use hdk::prelude::*;

#[derive(Clone, Serialize, Deserialize, Debug)]
pub enum RemoteSignal {
    MovieLoanHasBeenCreatedForYou(ActionHash),
}

> #[hdk_extern]

Reviewer: Yes, it's not allowed to error

> pub fn post_commit(actions: Vec<SignedActionHashed>) -> ExternResult<()> {

Reviewer: This example can't be run because the integrity types are missing?

Author: see preceding paragraph -- it references the integrity types defined in another doc.
